pkg/virtual/framework/types.go

```diff
@@ -30,10 +30,49 @@ import (
 // the request to the given VirtualWorkspace delegated APIServer.
 type RootPathResolverFunc func(urlPath string, context context.Context) (accepted bool, prefixToStrip string, completedContext context.Context)
 
+func (r RootPathResolverFunc) ResolveRootPath(urlPath string, context context.Context) (accepted bool, prefixToStrip string, completedContext context.Context) {
+	return r(urlPath, context)
+}
+
+var _ RootPathResolver = RootPathResolverFunc(nil)
+
+type RootPathResolver interface {
+	// ResolveRootPath returns whether the request should be accepted and served by a given VirtualWorkspace. If accepted,
+	// the prefixToStrip is the prefix that is in-front of the kube-like API surface (including `/clusters/<something>`). E.g. for
+	//
+	//   /services/initializingworkspaces/<initializer>/clusters/<something>/apis/workload.kcp.dev/v1alpha1/workloadclusters
+	//
+	// the prefixToStrip is `/services/initializingworkspaces/<initializer/clusters/<something>`. The returned context holds the APIDomainKey and
+	// the cluster name (e.g. `<something>` here).
+	ResolveRootPath(urlPath string, context context.Context) (accepted bool, prefixToStrip string, completedContext context.Context)
+}
+
 // ReadyFunc is the type of readiness check functions exposed by types
 // implementing the VtualWorkspace interface.
 type ReadyFunc func() error
 
+func (r ReadyFunc) IsReady() error {
+	return r()
+}
+
+var _ ReadyChecker = ReadyFunc(nil)
+
+type ReadyChecker interface {
+	IsReady() error
+}
+
+type Name string
+
+func (n Name) GetName() string {
+	return string(n)
+}
+
+var _ Namer = Name("")
+
+type Namer interface {
```

Comment:
Are all these individual interfaces really necessary ?
I'm not convinced about them making the code clearer.

Reply:
@sttts wdyt ?

Reply:
alternative suggestion, last commit: https://github.com/kcp-dev/kcp/pull/1328 (follow-up)

Reply:
about the interfaces, I tend to agree. These add a level of indirection (= complexity) without value.

Reply:
But this is clearly taste and not blocking here.

Reply:
I don't want to block the PR for this.

Reply:
Before these changes, every single VirtualWorkspace would hold the name, ready checker, root resolver, etc - and, every single one of them had to explicitly use these subordinate objects to implement functionality. Convince me copy-pasting this:

```go

func (vw *DynamicVirtualWorkspace) GetName() string {
	return vw.Name
}

func (vw *DynamicVirtualWorkspace) IsReady() error {
	return vw.Ready()
}

func (vw *DynamicVirtualWorkspace) ResolveRootPath(urlPath string, context context.Context) (accepted bool, prefixToStrip string, completedContext context.Context) {
	return vw.RootPathResolver(urlPath, context)
}

func (vw *DynamicVirtualWorkspace) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
	if vw.Authorizer != nil {
		return vw.Authorizer(ctx, a)
	}
	return authorizer.DecisionNoOpinion, "", nil
}
```

Is somehow better than composition?

Reply:
So IMHO the interfaces are already currently implicitly there, since we have e.g. `RootResolverFunc` but no `RootResolver`. We can remove the interface and make it less clear that `RootResolverFunc` is used as a component to compose `VirtualWorkspce` (implicit here is not bad necessarily) but IMHO writing clearly that:

1. a `VirtualWorkspace` needs to be a `RootResolver`
2. a `RootResolverFunc` is a `RootResolver`

Adds nothing new on top, just makes the current structure explicit & clear.